plugin/bind/setup.go

```diff
@@ -10,48 +11,101 @@ import (
 )
 
 func setup(c *caddy.Controller) error {
-	config := dnsserver.GetConfig(c)
 
+	config := dnsserver.GetConfig(c)
 	// addresses will be consolidated over all BIND directives available in that BlocServer
 	all := []string{}
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return plugin.Error("bind", fmt.Errorf("failed to get interfaces list"))
+	}
+
 	for c.Next() {
-		args := c.RemainingArgs()
-		if len(args) == 0 {
-			return plugin.Error("bind", fmt.Errorf("at least one address or interface name is expected"))
+		b, err := parse(c)
+		if err != nil {
+			return plugin.Error("bind", err)
 		}
 
-		ifaces, err := net.Interfaces()
+		ips, err := listIP(b.addrs, ifaces)
 		if err != nil {
-			return plugin.Error("bind", fmt.Errorf("failed to get interfaces list"))
-		}
-
-		var isIface bool
-		for _, arg := range args {
-			isIface = false
-			for _, iface := range ifaces {
-				if arg == iface.Name {
-					isIface = true
-					addrs, err := iface.Addrs()
-					if err != nil {
-						return plugin.Error("bind", fmt.Errorf("failed to get the IP addresses of the interface: %q", arg))
-					}
-					for _, addr := range addrs {
-						if ipnet, ok := addr.(*net.IPNet); ok {
-							if ipnet.IP.To4() != nil || (!ipnet.IP.IsLinkLocalMulticast() && !ipnet.IP.IsLinkLocalUnicast()) {
-								all = append(all, ipnet.IP.String())
-							}
+			return err
```

Comment:
can you move the plugin.Error calls from `parse` here? So only this function has those? Had to look twice if this was done correctly in the current code

Reply:
Or at least have a single function where plugin.Error is called, usually this is only done in `setup`

Reply:
Nice catch 👌 I moved them around to new functions and missed that duplicate plugin.Error call. Now I moved all plugin.Error from `parse` and `listIP` to `setup` function..